cpp/JobSystem.hpp

```diff
@@ -0,0 +1,175 @@
+#pragma once
+
+#include "Thread.hpp"
+#include "Functional.hpp"
+#include "Vector.hpp"
+#include "MPMCQueue.hpp"
+
+#include <atomic>
+
+namespace hsd
+{
+    struct Job;
+
+    using JobFn = void(*)(Job);
+    using Counter = std::atomic<size_t>;
+
+    struct Job
+    {
+        JobFn task;
+        void** data = nullptr;
+    };
+
+/*
+    template<u8 numargs, typename T>
+    struct JobData
+    {
+        
+    };
+*/
+
+    enum PRIO
+    {
+        HIGH = 0,
+        NORM,
+        LOW,
+
+        NUM_PRIORITIES = 3
+    };
+
+    namespace priv
+    {
+        class JobSystem
+        {
+        private:
+            hsd::Counter counter = {0};
+            std::atomic_bool running = {false};
+
+            MPMCQueue<Job> high_priority;
+            MPMCQueue<Job> normal_priority;
+            MPMCQueue<Job> low_priority;
+
+            hsd::vector<hsd::thread> threads;
+
+            void thread_function()
+            {
+                hsd::Job job;
+                while(running.load())
+                {
+                    if(high_priority.try_pop(job))
+                    {
+                        job.task(job);
+                    }
+                    else if(normal_priority.try_pop(job))
+                    {
+                        job.task(job);
+                    }
+                    else if(low_priority.try_pop(job))
+                    {
+                        job.task(job);
+                    }
+                }
+            }
+
+        public:
+            JobSystem() : high_priority(512), normal_priority(512), low_priority(512)
+            {
+                running.store(true);
+
+                auto numcores = hsd::thread::hardware_concurrency();
+                numcores = 4;
```

Comment:
What is this? You get the core count, yet you still make it 4

Reply:
Because the function is hardcoded to return 0. I was considering doing some template shenanigans for this, but decided against it.